The Perceptron
Receives inputs, multiplies them by some weight, and then passes them into an activation function to produce an output

In [1]:
from sklearn.datasets import load_breast_cancer #Binary classification dataset
#Complete dataset- https://goo.gl/U2Uwz2
cancer=load_breast_cancer()

In [2]:
cancer.keys() #its like a dictionary

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])

In [12]:
cancer['feature_names']

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

In [13]:
cancer['DESCR']

'Breast Cancer Wisconsin (Diagnostic) Database\n=============================================\n\nNotes\n-----\nData Set Characteristics:\n    :Number of Instances: 569\n\n    :Number of Attributes: 30 numeric, predictive attributes and the class\n\n    :Attribute Information:\n        - radius (mean of distances from center to points on the perimeter)\n        - texture (standard deviation of gray-scale values)\n        - perimeter\n        - area\n        - smoothness (local variation in radius lengths)\n        - compactness (perimeter^2 / area - 1.0)\n        - concavity (severity of concave portions of the contour)\n        - concave points (number of concave portions of the contour)\n        - symmetry \n        - fractal dimension ("coastline approximation" - 1)\n\n        The mean, standard error, and "worst" or largest (mean of the three\n        largest values) of these features were computed for each image,\n        resulting in 30 features.  For instance, field 3 is Mean Rad

In [14]:
type(cancer['data'])

numpy.ndarray

In [15]:
import sys
sys.getsizeof(cancer)

256

In [16]:
cancer['data'].shape

(569, 30)

In [3]:
#setting up our data- x and y variables
x=cancer['data']
y=cancer['target']

In [4]:
#train-test split
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test= train_test_split(x,y,test_size=0.33, random_state=42) #proportion of test data and random seed

In [5]:
#data pre-processing
# Multi-layer Perceptron is sensitive to feature scaling
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
scaler.fit(x_train)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [6]:
# Now apply the transformations to the data:
x_train = scaler.transform(x_train)
x_test = scaler.transform(x_test)

In [7]:
from sklearn.neural_network import MLPClassifier

In [9]:
# for simplicity we will choose 3 layers
#MLPClassifier trains using Backpropagation (Gradient Descent)
mlp=MLPClassifier(hidden_layer_sizes= (30,30,30), max_iter=500, alpha=0.0001,
                     solver='sgd', verbose=10,  random_state=21,tol=0.000000001) #three hidden layers with 30 units each
mlp

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(30, 30, 30), learning_rate='constant',
       learning_rate_init=0.001, max_iter=500, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=21, shuffle=True,
       solver='sgd', tol=1e-09, validation_fraction=0.1, verbose=10,
       warm_start=False)

class sklearn.neural_network.MLPClassifier(hidden_layer_sizes=(100, ), activation=’relu’, solver=’adam’, alpha=0.0001, batch_size=’auto’, learning_rate=’constant’, learning_rate_init=0.001, power_t=0.5, max_iter=200, shuffle=True, random_state=None, tol=0.0001, verbose=False, warm_start=False, momentum=0.9, nesterovs_momentum=True, early_stopping=False, validation_fraction=0.1, beta_1=0.9, beta_2=0.999, epsilon=1e-08)

In [24]:
#fit the training data to our model
mlp.fit(x_train,y_train) #the model will accept a bunch of default values for other parameters

Iteration 1, loss = 0.72075737
Iteration 2, loss = 0.71877200
Iteration 3, loss = 0.71576550
Iteration 4, loss = 0.71196943
Iteration 5, loss = 0.70759033
Iteration 6, loss = 0.70288249
Iteration 7, loss = 0.69782504
Iteration 8, loss = 0.69262888
Iteration 9, loss = 0.68734754
Iteration 10, loss = 0.68212515
Iteration 11, loss = 0.67679281
Iteration 12, loss = 0.67147887
Iteration 13, loss = 0.66650535
Iteration 14, loss = 0.66133396
Iteration 15, loss = 0.65646619
Iteration 16, loss = 0.65165437
Iteration 17, loss = 0.64676365
Iteration 18, loss = 0.64213597
Iteration 19, loss = 0.63759865
Iteration 20, loss = 0.63318775
Iteration 21, loss = 0.62883557
Iteration 22, loss = 0.62467666
Iteration 23, loss = 0.62050001
Iteration 24, loss = 0.61642903
Iteration 25, loss = 0.61241744
Iteration 26, loss = 0.60844988
Iteration 27, loss = 0.60461326
Iteration 28, loss = 0.60081861
Iteration 29, loss = 0.59703917
Iteration 30, loss = 0.59325898
Iteration 31, loss = 0.58964278
Iteration 32, los

Iteration 287, loss = 0.10263962
Iteration 288, loss = 0.10237687
Iteration 289, loss = 0.10210508
Iteration 290, loss = 0.10181180
Iteration 291, loss = 0.10156042
Iteration 292, loss = 0.10128266
Iteration 293, loss = 0.10102878
Iteration 294, loss = 0.10076204
Iteration 295, loss = 0.10048057
Iteration 296, loss = 0.10022400
Iteration 297, loss = 0.09996555
Iteration 298, loss = 0.09971431
Iteration 299, loss = 0.09944688
Iteration 300, loss = 0.09921263
Iteration 301, loss = 0.09896147
Iteration 302, loss = 0.09868431
Iteration 303, loss = 0.09845762
Iteration 304, loss = 0.09818905
Iteration 305, loss = 0.09795503
Iteration 306, loss = 0.09771231
Iteration 307, loss = 0.09747171
Iteration 308, loss = 0.09722014
Iteration 309, loss = 0.09700061
Iteration 310, loss = 0.09675869
Iteration 311, loss = 0.09650620
Iteration 312, loss = 0.09628463
Iteration 313, loss = 0.09604858
Iteration 314, loss = 0.09581305
Iteration 315, loss = 0.09558239
Iteration 316, loss = 0.09535229
Iteration 

C:\Users\Gautam\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:564: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(30, 30, 30), learning_rate='constant',
       learning_rate_init=0.001, max_iter=500, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=21, shuffle=True,
       solver='sgd', tol=1e-09, validation_fraction=0.1, verbose=10,
       warm_start=False)

In [20]:
 predictions=mlp.predict(x_test) #run predictions on the test data

In [22]:
#use sklearn's built in metrics for confusion matrix
from sklearn.metrics import classification_report,confusion_matrix
print(confusion_matrix(y_test,predictions))

[[43  1]
 [ 2 97]]


In [23]:
print(classification_report(y_test,predictions))

             precision    recall  f1-score   support

          0       0.96      0.98      0.97        44
          1       0.99      0.98      0.98        99

avg / total       0.98      0.98      0.98       143

